python/cudf_polars/cudf_polars/experimental/rapidsmpf/shuffle.py

```diff
@@ -0,0 +1,285 @@
+
+from __future__ import annotations
+
+import threading
+from typing import TYPE_CHECKING, Any, Literal
+
+from rapidsmpf.communicator.single import new_communicator
+from rapidsmpf.config import Options, get_environment_variables
+from rapidsmpf.integrations.cudf.partition import (
+    partition_and_pack as py_partition_and_pack,
+    unpack_and_concat as py_unpack_and_concat,
+)
+from rapidsmpf.progress_thread import ProgressThread
+from rapidsmpf.shuffler import Shuffler
+from rapidsmpf.streaming.core.message import Message
+from rapidsmpf.streaming.core.node import define_py_node
+from rapidsmpf.streaming.cudf.table_chunk import TableChunk
+
+from rmm.pylibrmm.stream import DEFAULT_STREAM
+
+from cudf_polars.dsl.expr import Col
+from cudf_polars.experimental.rapidsmpf.dispatch import (
+    generate_ir_sub_network,
+)
+from cudf_polars.experimental.rapidsmpf.nodes import shutdown_on_error
+from cudf_polars.experimental.rapidsmpf.utils import ChannelManager
+from cudf_polars.experimental.shuffle import Shuffle
+
+if TYPE_CHECKING:
+    from types import TracebackType
+
+    from rapidsmpf.streaming.core.context import Context
+
+    import pylibcudf as plc
+    from rmm.pylibrmm.stream import Stream
+
+    from cudf_polars.dsl.ir import IR, IRExecutionContext
+    from cudf_polars.experimental.rapidsmpf.core import SubNetGenerator
+    from cudf_polars.experimental.rapidsmpf.utils import ChannelPair
+
+
+# TODO: This implementation only supports a single GPU for now.
+#       Multi-GPU support will require a distinct GlobalShuffle
+#       context manager, and updated _shuffle_id_vacancy logic.
+
+
+# Set of available shuffle IDs
+_shuffle_id_vacancy: set[int] = set(range(Shuffler.max_concurrent_shuffles))
+_shuffle_id_vacancy_lock: threading.Lock = threading.Lock()
+
+
+def _get_new_shuffle_id() -> int:
+    with _shuffle_id_vacancy_lock:
+        if not _shuffle_id_vacancy:
+            raise ValueError(
+                f"Cannot shuffle more than {Shuffler.max_concurrent_shuffles} "
+                "times in a single query."
+            )
+
+        return _shuffle_id_vacancy.pop()
+
+
+def _release_shuffle_id(op_id: int) -> None:
+    """Release a shuffle ID back to the vacancy set."""
+    with _shuffle_id_vacancy_lock:
+        _shuffle_id_vacancy.add(op_id)
+
+
+class LocalShuffle:
+    """
+    Local shuffle instance context manager.
+
+    Parameters
+    ----------
+    context: Context
+        The streaming context.
+    num_partitions: int
+        The number of partitions to shuffle into.
+    columns_to_hash: tuple[int, ...]
+        The columns to hash.
+    stream: Stream
+        The stream to use for the shuffle.
+    """
+
+    def __init__(
+        self,
+        context: Context,
+        num_partitions: int,
+        columns_to_hash: tuple[int, ...],
+        stream: Stream,
+    ):
+        self.context = context
+        self.br = context.br()
+        self.op_id = _get_new_shuffle_id()
```

Comment:
Probably not blocking, but I have a slight preference to anything other than attribute assignment (`self.context = context, self.num_partitions=num_partitions, ...`) in the `__enter__`.